### DISCUSSION TITLE
Subnavigation query

### DISCUSSION THREAD

**Discussion by Diewy** [verbb/navigation#368] [Q&A]
We have the following navigation structure, as seen in the documention on Eager-Loading:

```
- Node 1
    - Node 1-1
    - Node 1-2
    - Node 1-3
    - Node 1-4
- Node 2
- Node 3
    - Node 3-1
    - Node 3-2
    - Node 3-3
```

Now, on each page of the second level, we would like to include a subnavigation. Including all nodes of that level.

So the result on each page should be:
Node 1: No subnavigation.
Node 1-1, 1-2, 1-3, and 1-4:
```
- Node 1-1
- Node 1-2
- Node 1-3
- Node 1-4
```
Node 2: No subnavigation
Node 3: No subnavigation
Node 3-1, 3-2, and 3-3:
```
- Node 3-1
- Node 3-2
- Node 3-3
```

I have it working with the following queries in my template:

```twig
{# Find corresponding navigation node in main navigation #}
{% set navigationNode = craft
      .navigation
      .nodes()
      .handle('main')
      .type('craft\\elements\\Entry')
      .elementId(entry.id)
      .with(['parent.children'])
      .one() %}

{# Get all the children of the parent navigation node #}
{% set parentNavigationNode = navigationNode.parent.children %}

 <ul>
      {% for sibling in parentNavigationNode %}
            <li>{{ sibling.getLink() }}</li>
      {% endfor %}
</ul>
```

But I was wondering if there would be an easier way to accomplish this.
Or, if there would be a way to optimize those queries.

Any suggestions or ideas are welcome.

Thanks!

**Comment by engram-design** [comment]
That looks fine to me! This seems unique in that you want to show the siblings, not the children? If you're on Node 1, technically the children of that current page are Node 1-1, 1-2, 1-3, and 1-4. But instead, when on Node 1-1 you want to view the siblings of that page.

In that case, there's no way around the approach you're doing here where you need to bubble up to the parent to get the children (siblings) to eager-load.

I was going to suggest querying the siblings:

```
{% set nodes = craft.navigation.nodes()
  .siblingOf(currentNode)
  .all() %}
```

But that will produce incorrect results when you're at the top-level - it'll report back the Node 2, etc nodes.

**Comment by engram-design** [comment]
Updated in [4.0.0-beta.1](https://github.com/verbb/navigation/releases/tag/4.0.0-beta.1). 

We've added a `craft.navigation.context()` API which should hopefully make things a bit easier.

```twig
{% set ctx = craft.navigation.context('main') %}

{% if ctx.parent() %}
  <ul>
    {% for node in ctx.siblings() %}
      <li class="{{ node.getCurrent() ? 'is-current' }}">
        <a href="{{ node.url }}">{{ node.title }}</a>
      </li>
    {% endfor %}
  </ul>
{% endif %}
```

`ctx.siblings()` returns same-level nodes under the shared parent. The `{% if ctx.parent() %}` guard avoids showing a sub-nav on top-level pages (where siblings would otherwise be all level-1 nodes).

Also available: `branch()`, `children()`, `ancestors()`, and GraphQL `navigationContext { siblings { title url } }`.

Be sure to check out the updated [docs](https://verbb.io/craft-plugins/navigation/docs/v4/get-started/installation-setup).
